### PR TITLE
refactor!: Move and Grid Density finders to cpp

### PR DESCRIPTION
**Pull request by paulgessinger**
This PR moves 

- `GaussianGridTrackDensity`
- `GridDensityVertexFinder`

to cpp files. It has to refactor the interface a bit to make this possible, mostly, the grid sizes become runtime arguments instead of compile-time parameters. This then also changes the internal grid eigen objects to be dynamic, which I think should be fine because we essentially only do lookups on them.

Part of:
- #2842

Blocked by:
- #2971

### CODE REVIEW

**Comment by paulgessinger** [comment]
Now this has some changes that we might want to discuss.

File: Core/src/Vertexing/GaussianGridTrackDensity.cpp [hunk 1]
```diff
@@ -93,71 +92,61 @@ Acts::GaussianGridTrackDensity<mainGridSize, trkGridSize>::addTrack(
...
   return trackGrid;
 }
 
-template <int mainGridSize, int trkGridSize>
-Acts::Result<float>
-Acts::GaussianGridTrackDensity<mainGridSize, trkGridSize>::estimateSeedWidth(
+Result<float> GaussianGridTrackDensity::estimateSeedWidth(
     MainGridVector& mainGrid, float maxZ) const {
-  if (mainGrid == MainGridVector::Zero()) {
+  if (mainGrid.isZero(0)) {
```
Comment (andiwand) [inline Core/src/Vertexing/GaussianGridTrackDensity.cpp:145]:
```suggestion
  if (mainGrid.isZero()) {
```

File: Core/src/Vertexing/GridDensityVertexFinder.cpp [hunk 2]
```diff
@@ -58,7 +61,7 @@ auto Acts::GridDensityVertexFinder<mainGridSize, trkGridSize>::find(
 
   double z = 0;
   double width = 0;
-  if (state.mainGrid != MainGridVector::Zero()) {
+  if (!state.mainGrid.isZero(0)) {
```
Comment (andiwand) [inline Core/src/Vertexing/GridDensityVertexFinder.cpp:64]:
```suggestion
  if (!state.mainGrid.isZero()) {
```

**Review by andiwand — APPROVED** [review]
making this is runtime config makes total sense to me. no clue why this was a template parameter in the first place

File: Core/include/Acts/Vertexing/GaussianGridTrackDensity.hpp [hunk 3]
```diff
@@ -21,32 +21,43 @@ namespace Acts {
...
+      }
+      if (mainGridSize < trkGridSize) {
+        throw std::runtime_error(
+            "GaussianGridTrackDensity: mainGridSize has to be bigger than "
+            "trkGridSize!");
+      }
     }
+
+    int mainGridSize;
+    int trkGridSize;
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/Vertexing/GaussianGridTrackDensity.hpp:59]: There's an explicit constructor that always initializes these, and no default constructor. Is this needed?

[the remark this answers is not in the document]
Comment (andiwand) [reply Core/include/Acts/Vertexing/GaussianGridTrackDensity.hpp:59]: no strong requirement - I just try to initialize everything so if somebody adds a new constructor and changes the existing one we never end up with uninitialized variables.

File: Core/src/Vertexing/GaussianGridTrackDensity.cpp (hunk 1 above)
Comment (paulgessinger) [reply Core/src/Vertexing/GaussianGridTrackDensity.cpp:145]: I faintly remember this was not doing the right thing without explicit precision, but I'll try.

File: Core/include/Acts/Vertexing/GaussianGridTrackDensity.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/Vertexing/GaussianGridTrackDensity.hpp:59]: The other variables (not part of the changeset) are also not initialized. In practice this should not make a difference here.

File: Core/include/Acts/Vertexing/GridDensityVertexFinder.hpp [hunk 4]
```diff
@@ -29,33 +29,23 @@ namespace Acts {
...
 
   /// @brief The Config struct
   struct Config {
-    ///@param zMinMax min and max z value of big z-axis grid
-    Config(float zMinMax = 100)
-        : gridDensity(typename GridDensity::Config(zMinMax)) {}
+    Config(float zMinMax_ = 100, int mainGridSize_ = 2000,
+           int trkGridSize_ = 15)
+        : Config{GaussianGridTrackDensity::Config{zMinMax_, mainGridSize_,
+                                                  trkGridSize_}} {}
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/Vertexing/GridDensityVertexFinder.hpp:42]: There's a constructor below that does this. I'll try removing this one.

File: Core/src/Vertexing/GaussianGridTrackDensity.cpp (hunk 1 above)
Comment (andiwand) [reply Core/src/Vertexing/GaussianGridTrackDensity.cpp:145]: oh right `isZero()` is actually checking against an epsilong. w-t-f

Comment (paulgessinger) [reply Core/src/Vertexing/GaussianGridTrackDensity.cpp]: Seems to be ok.

**Review by andiwand — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [5eb39c3c709f8c38805115fd530b8258dc2c3850]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567084)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567083)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567082)||
|:green_circle: | [test_ActsConversionWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567081)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567080)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567079)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567078)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567077)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567076)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567075)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567074)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567073)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567072)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567071)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567070)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567069)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567068)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567067)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567066)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567065)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567064)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567063)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567062)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567061)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36567060)||

**Pull request merged**